[Table of Contents](../../index.ipynb)

# FRC Analytics with Python - Session 09
# Tabular Data
**Last Updated: 20 December 2020**

## I. Introduction
Tabular data is data that is organized into rows and columns. It's everywhere -- in newspapers and magazines, on websites, in computer databases, and in Microsoft Excel files. Evaluating and manipulating tabular data is an essential skill for any analyst.

So far we've experimented with lists, tuples, and dictionaries. These data structures are indispensable, but they are not optimal for working with tables. In this session we will review several techniques and tools for working with tabular data.

## II. Comma Separated Value (CSV) Files
### A. CSV File Structure
Tabular data is often stored in comma separated value (CSV) files. CSV files are text files that use commas and newline characters (I'll explain what those are shortly) to organize the contents of the file into a table. Let's look at an example. The `space.csv` file contains information on 4,324 space launches, starting with the launch of the the Sputnik spacecraft by the Soviet Union in 1957. The dataset is available on the [Kaggle website](https://www.kaggle.com/agirlcoding/all-space-missions-from-1957). The Python code below opens the file and displays the first five lines.

In [2]:
# Open a text file and print the first five lines
# Don't worry if you don't understand all of this code.
with open("space.csv", "rt", encoding="UTF-8") as csv_file:
    for row in range(5):
        print(csv_file.readline())

Company Name,Location,Datum,Detail,Status Rocket,Rocket,Status Mission

SpaceX,"LC-39A, Kennedy Space Center, Florida, USA","Fri Aug 07, 2020 05:12 UTC",Falcon 9 Block 5 | Starlink V1 L9 & BlackSky,StatusActive,50,Success

CASC,"Site 9401 (SLS-2), Jiuquan Satellite Launch Center, China","Thu Aug 06, 2020 04:01 UTC",Long March 2D | Gaofen-9 04 & Q-SAT,StatusActive,29.75,Success

SpaceX,"Pad A, Boca Chica, Texas, USA","Tue Aug 04, 2020 23:57 UTC",Starship Prototype | 150 Meter Hop,StatusActive,,Success

Roscosmos,"Site 200/39, Baikonur Cosmodrome, Kazakhstan","Thu Jul 30, 2020 21:25 UTC",Proton-M/Briz-M | Ekspress-80 & Ekspress-103,StatusActive,65,Success



The first row of text contains the column headings, with each column separated by a comma. The subsequent rows contain the data, with one row for each space launch. The rows are separated from each other with a newline character.

Commas are also used for separation in the data rows, but it's a bit difficult to keep track of what text belongs to which column. Many of the data values contain commas inside the data. For example, the second column in the first row contains the value *"LC-39A, Kennedy Space Center, Florida, USA"*. The commas within quotation marks are part of the data and are not used for column separation.

CSV files are popular because they are simple to create and can be opened and read with any text editor. Still, the content can be tedious to read. All of the values within a row are smashed together and the columns in the data rows do not line up with the column headers.

### B. Python CSV Module
We saw earlier how we can use a built-in Python function like `open()` to read data from a CSV file on disk, but the results can be difficult to read. The Python Standard Library has a `csv` module that makes things a little better.

In [3]:
import csv
space_csv = []
with open("space.csv", "rt", encoding="UTF-8") as csv_file:
    reader = csv.reader(csv_file)
    for row in reader:
        space_csv.append(row)

In [4]:
space_csv[:3]

[['Company Name',
  'Location',
  'Datum',
  'Detail',
  'Status Rocket',
  'Rocket',
  'Status Mission'],
 ['SpaceX',
  'LC-39A, Kennedy Space Center, Florida, USA',
  'Fri Aug 07, 2020 05:12 UTC',
  'Falcon 9 Block 5 | Starlink V1 L9 & BlackSky',
  'StatusActive',
  '50',
  'Success'],
 ['CASC',
  'Site 9401 (SLS-2), Jiuquan Satellite Launch Center, China',
  'Thu Aug 06, 2020 04:01 UTC',
  'Long March 2D | Gaofen-9 04 & Q-SAT',
  'StatusActive',
  '29.75',
  'Success']]

The `csv` module converts every row of the CSV file to a Python list. We appended every row to an outer list, to create a lists of lists, or a nested list. Each value is now put on its own row. We can even extract individual values from the nested list. For example, to get the third element of the third row:

In [5]:
space_csv[2][2]

'Thu Aug 06, 2020 04:01 UTC'

But what if we wanted to figure out how many space launches occurred in China on Wednesdays since the year 2000? That would require us to write several lines of code to read through all of the rows of data and count the applicable launches. Fortunately Python has a better tool for working with tabular data.

## III. Pandas Package
The *Pandas* package is an excellent tool for working with tabular data. Pandas is not included by default when installing Python, but it can easily be installed by running the command `conda install pandas`.

Pandas has so many features that it would take a book to explain them all. This session omits several important Pandas topics, such as timeseries data and multi-level indices. Students are encouraged to become familiar with the official Pandas documentation, which is located here: https://pandas.pydata.org/pandas-docs/stable/index.html. The remaining sections of this notebook contain links to applicable portions of the documentation, which provide additional information on each topic.

### A. Getting Started
 Let's see how our space data looks when we use Pandas to view it.

In [6]:
import pandas as pd
space_df = pd.read_csv("space.csv", thousands=",")
space_df.head()

Company Name                                           Location  \
0       SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
1         CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2       SpaceX                      Pad A, Boca Chica, Texas, USA   
3    Roscosmos       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4          ULA           SLC-41, Cape Canaveral AFS, Florida, USA   

                        Datum                                        Detail  \
0  Fri Aug 07, 2020 05:12 UTC  Falcon 9 Block 5 | Starlink V1 L9 & BlackSky   
1  Thu Aug 06, 2020 04:01 UTC           Long March 2D | Gaofen-9 04 & Q-SAT   
2  Tue Aug 04, 2020 23:57 UTC            Starship Prototype | 150 Meter Hop   
3  Thu Jul 30, 2020 21:25 UTC  Proton-M/Briz-M | Ekspress-80 & Ekspress-103   
4  Thu Jul 30, 2020 11:50 UTC                    Atlas V 541 | Perseverance   

  Status Rocket  Rocket Status Mission  
0  StatusActive   50.00        Success  
1  StatusActive   29.75        Success  
2  StatusActive     NaN        Success  
3  StatusActive   65.00        Success  
4  StatusActive  145.00        Success

Now that is much better. All of the data lines up with the column headers. Pandas even adds row numbers and shades alternate rows to make everything easy to read. And we did everything in three short lines of code:
* The first line imports the pandas module and renames it `pd`.
* The next line reads the CSV file and creates a `DataFrame` object.
* The final line displays the `DataFrame` object. The `.head()` method causes only the first five lines to be displayed. You can customize the number of lines displayed with .head() by putting the number in the parenthesis.

By the way, the package isn't named *Pandas* because the developers really like pandas (but who doesn't like pandas?). *Pandas* is short for *panel data*. Panel data is common in the social sciences. It is multi-dimensional data on on multiple entities, with measurements taken at several points in time. For example, suppose we're conducting a study on family income over time. We might collect multiple pieces of data on each family, such as income, number of children, education level of parents, age of parents, whether they own their home, etc. If we collect such information on 500 families, and then update the information every year for five years, we have panel data.

Python's `len()` function can be used with dataframes to get the number of rows.

In [7]:
# Using len() with DataFrames
len(space_df)

4324

Pandas `DataFrame` objects have a `shape` attribute that contains a two-element tuple (immutable lists). The first element is the number of rows and the second is the number of columns.

In [8]:
print("Number of rows and columns:", space_df.shape)
print("Just the number of columns:", space_df.shape[1])

Number of rows and columns: (4324, 7)
Just the number of columns: 7


##### More Information
* [Intro to Dataframes](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html)
* [Read and Write Tabular Data](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/02_read_write.html)
* [10 Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html)
* [Essential Functionality](https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html)

### B. Easy Pandas Exercises

**Ex. III.1** The `.head()` method will accept an integer argument that represents the number of rows to display. Display the first eight rows of the dataframe.

In [9]:
# Ex. III.1
space_df.head(8)

Company Name                                           Location  \
0       SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
1         CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2       SpaceX                      Pad A, Boca Chica, Texas, USA   
3    Roscosmos       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4          ULA           SLC-41, Cape Canaveral AFS, Florida, USA   
5         CASC       LC-9, Taiyuan Satellite Launch Center, China   
6    Roscosmos         Site 31/6, Baikonur Cosmodrome, Kazakhstan   
7         CASC    LC-101, Wenchang Satellite Launch Center, China   

                        Datum  \
0  Fri Aug 07, 2020 05:12 UTC   
1  Thu Aug 06, 2020 04:01 UTC   
2  Tue Aug 04, 2020 23:57 UTC   
3  Thu Jul 30, 2020 21:25 UTC   
4  Thu Jul 30, 2020 11:50 UTC   
5  Sat Jul 25, 2020 03:13 UTC   
6  Thu Jul 23, 2020 14:26 UTC   
7  Thu Jul 23, 2020 04:41 UTC   

                                              Detail Status Rocket  Rocket  \
0       Falcon 9 Block 5 | Starlink V1 L9 & BlackSky  StatusActive   50.00   
1                Long March 2D | Gaofen-9 04 & Q-SAT  StatusActive   29.75   
2                 Starship Prototype | 150 Meter Hop  StatusActive     NaN   
3       Proton-M/Briz-M | Ekspress-80 & Ekspress-103  StatusActive   65.00   
4                         Atlas V 541 | Perseverance  StatusActive  145.00   
5  Long March 4B | Ziyuan-3 03, Apocalypse-10 & N...  StatusActive   64.68   
6                        Soyuz 2.1a | Progress MS-15  StatusActive   48.50   
7                           Long March 5 | Tianwen-1  StatusActive     NaN   

  Status Mission  
0        Success  
1        Success  
2        Success  
3        Success  
4        Success  
5        Success  
6        Success  
7        Success

**Ex. III.2** There is also a `tail()` method that will display the last few rows of a dataframe. Display the last 4 rows of the `space_df` dataframe.

In [10]:
# Ex. III.2
space_df.tail(4)

Company Name                                   Location  \
4320         AMBA   LC-26A, Cape Canaveral AFS, Florida, USA   
4321      US Navy   LC-18A, Cape Canaveral AFS, Florida, USA   
4322    RVSN USSR  Site 1/5, Baikonur Cosmodrome, Kazakhstan   
4323    RVSN USSR  Site 1/5, Baikonur Cosmodrome, Kazakhstan   

                           Datum                      Detail  Status Rocket  \
4320  Sat Feb 01, 1958 03:48 UTC         Juno I | Explorer 1  StatusRetired   
4321  Fri Dec 06, 1957 16:44 UTC     Vanguard | Vanguard TV3  StatusRetired   
4322  Sun Nov 03, 1957 02:30 UTC  Sputnik 8K71PS | Sputnik-2  StatusRetired   
4323  Fri Oct 04, 1957 19:28 UTC  Sputnik 8K71PS | Sputnik-1  StatusRetired   

      Rocket Status Mission  
4320     NaN        Success  
4321     NaN        Failure  
4322     NaN        Success  
4323     NaN        Success

### C. Pandas Data Types
Pandas provides two different types of data structures: `Series` and `DataFrame`. 

#### DataFrame
What type of data structure is the `space_df` object?

In [10]:
# What type of object is space_df?
type(space_df)

pandas.core.frame.DataFrame

The `space_df` object is an object of type [`DataFrame`](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html#dataframe). Most `DataFrame` objects are two-dimensional with rows and columns. `DataFrames` can be modified to contain data with three or more dimensions, such as panel data, but we won't bother with that in this course.

Pay attention to the capitalization of `DataFrame` (capital F). We're using the word "dataframe" a couple different ways. A dataframe is a two-dimensional data structure that can be found in Python, [R](https://en.wikipedia.org/wiki/R_(programming_language) (a language for statistical analysis), and [Julia](https://en.wikipedia.org/wiki/Julia_(programming_language) (a relatively new language that is good for numerical analysis). A `DataFrame`, on the other hand, is a Python data type provided by the *Pandas* package.

#### Series
The following code extracts a single column from the `space_df` dataframe and displays its type.

In [11]:
# pandas.Series data type
datum_series = space_df.Datum.head(6)  # Extract a single column(named Datum in this case) and display top six rows
print(datum_series)
type(datum_series)

0    Fri Aug 07, 2020 05:12 UTC
1    Thu Aug 06, 2020 04:01 UTC
2    Tue Aug 04, 2020 23:57 UTC
3    Thu Jul 30, 2020 21:25 UTC
4    Thu Jul 30, 2020 11:50 UTC
5    Sat Jul 25, 2020 03:13 UTC
Name: Datum, dtype: object


pandas.core.series.Series

See how we extracted a single column from the `DataFrame` by appending a period and its name to the name of the `DataFrame`? Extracting a single column results in a Pandas [`Series`](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html#series) object. A `Series` is similar to a list, but with some differences:
* Unlike a Python list, all elements of a `Series` must have the same data type. The *Datum* column's type is *object*, which is the type Pandas uses for strings.
* The contents of a Pandas `Series` are stored in memory more efficiently than lists. Because of this, calculations on `Series` objects are often faster than equivalent calculations on lists.

We will mostly use `DataFrame` objects instead of `Series` objects. But it's important to know what `Series` objects are because that's what we'll end up with whenever we extract a single column from a dataframe.

##### More Information
* [Pandas Data Structures](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html)

### D. Selecting Data within Pandas Dataframes
Pandas provides an immense number of ways to select and extract data from a dataframe, many more than can be covered in this class. Check out the [official Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html) to see for yourself. The Pandas documentation often refers to selecting data as *indexing*, which may seem strange at first. Think about it like this - to extract the fourth item from a list called `mylist`, we put the number 3 in square brackets: `fourth_num = mylist[3]`. The number 3 is the index position of the lists fourth element of the list. We're selecting a value from the list by passing an index to the list.

#### 1. Selecting a Single Column
Select a single column by using a dictionary-style notation. Place the column name in quotes and square brackets. This technique works for all columns regardless of their name.

In [12]:
# Extracting a single column
space_df["Status Mission"]

0       Success
1       Success
2       Success
3       Success
4       Success
         ...   
4319    Failure
4320    Success
4321    Failure
4322    Success
4323    Success
Name: Status Mission, Length: 4324, dtype: object

The dictionary style also works with a string variable that contains the column name.

In [12]:
col_var = "Status Rocket"
space_df[col_var]

0        StatusActive
1        StatusActive
2        StatusActive
3        StatusActive
4        StatusActive
            ...      
4319    StatusRetired
4320    StatusRetired
4321    StatusRetired
4322    StatusRetired
4323    StatusRetired
Name: Status Rocket, Length: 4324, dtype: object

#### 2. Selecting Part of a Single Column
We can select one or more rows from a single column the same way we select portions of a list.

In [13]:
space_df.Detail[100:105] # Detail is the name of the column

100                                  H-IIB | HTV-8
101        Long March 3B/YZ-1 | BeiDou-3 M23 & M24
102              Long March 11 | Zhuhai-1 Group 03
103    Long March 4B | Ziyuan-2D, BNU-1 & Taurus-1
104                   Kuaizhou 1A | KX-09 & Others
Name: Detail, dtype: object

#### 3. Selecting Multiple Columns
Multiple columns can be selected by passing a list of column names within square brackets. We can even change the column order.

In [14]:
space_df[["Location", "Company Name", "Status Mission"]] 
# Notice how with more than 1 specified column name, you need 2 sets of square brackets

Location Company Name  \
0            LC-39A, Kennedy Space Center, Florida, USA       SpaceX   
1     Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...         CASC   
2                         Pad A, Boca Chica, Texas, USA       SpaceX   
3          Site 200/39, Baikonur Cosmodrome, Kazakhstan    Roscosmos   
4              SLC-41, Cape Canaveral AFS, Florida, USA          ULA   
...                                                 ...          ...   
4319           LC-18A, Cape Canaveral AFS, Florida, USA      US Navy   
4320           LC-26A, Cape Canaveral AFS, Florida, USA         AMBA   
4321           LC-18A, Cape Canaveral AFS, Florida, USA      US Navy   
4322          Site 1/5, Baikonur Cosmodrome, Kazakhstan    RVSN USSR   
4323          Site 1/5, Baikonur Cosmodrome, Kazakhstan    RVSN USSR   

     Status Mission  
0           Success  
1           Success  
2           Success  
3           Success  
4           Success  
...             ...  
4319        Failure  
4320        Success  
4321        Failure  
4322        Success  
4323        Success  

[4324 rows x 3 columns]

#### 4. Selecting Rows
You might be tempted to select a row from a `DataFrame` the same way we select an element from a list. Resist that temptation, it won't work. Use `.loc[]` instead.

In [15]:
# Selecting part of a DataFrame with the `.loc()` function
space_df.loc[0:3, "Company Name":"Datum"]

Company Name                                           Location  \
0       SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
1         CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2       SpaceX                      Pad A, Boca Chica, Texas, USA   
3    Roscosmos       Site 200/39, Baikonur Cosmodrome, Kazakhstan   

                        Datum  
0  Fri Aug 07, 2020 05:12 UTC  
1  Thu Aug 06, 2020 04:01 UTC  
2  Tue Aug 04, 2020 23:57 UTC  
3  Thu Jul 30, 2020 21:25 UTC

To use the `.loc` function, pass two elements within square brackets, separated by a comma. The first element specifies what rows are selected, and the second specifies what columns are selected. List-style slice notation can be used to select ranges of rows and columns. In the example above, we selected rows 0 through 3 and columns "Company Name" through "Datum".

One difference between Pandas dataframe slice notation differs and Python list notation is that for Python lists, the slice does NOT return the final element. For example:

In [16]:
# Python list slices include all elements up to but NOT including the final
# element in the slice
# Will return 3 list items
tens = [0, 10, 20, 30, 40, 50, 60]
tens[0:3]  # The fourth element, 30, is not returned.

[0, 10, 20]

In [17]:
# Slices in DataFrames will include the final element of the slice
# Returns 4 rows and 4 columns
space_df.loc[0:3, "Detail":"Status Mission"]

Detail Status Rocket  Rocket  \
0  Falcon 9 Block 5 | Starlink V1 L9 & BlackSky  StatusActive   50.00   
1           Long March 2D | Gaofen-9 04 & Q-SAT  StatusActive   29.75   
2            Starship Prototype | 150 Meter Hop  StatusActive     NaN   
3  Proton-M/Briz-M | Ekspress-80 & Ekspress-103  StatusActive   65.00   

  Status Mission  
0        Success  
1        Success  
2        Success  
3        Success

Rows and columns need not be contiguous. We can pass in lists of row indices and column names.

In [18]:
space_df.loc[[100, 200, 300, 400], ["Detail", "Datum", "Rocket"]]

Detail                       Datum  \
100                             H-IIB | HTV-8  Tue Sep 24, 2019 16:05 UTC   
200  H-IIA 202 | Ibuki 2, KhalifaSat & Others  Mon Oct 29, 2018 04:08 UTC   
300                  Soyuz 2.1b | Cosmos 2524  Sat Dec 02, 2017 10:43 UTC   
400            Long March 5/YZ-2 | Shijian-17  Thu Nov 03, 2016 12:42 UTC   

     Rocket  
100   112.5  
200    90.0  
300    35.0  
400     NaN

#####  More Information
* [Selecting a Subset of a Dataframe (short intro)](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/03_subset_data.html)

### E. Pandas Indexing Exercises

#### Ex III.3
Display the datums and company names for rows 1318, 2976, and 4131.

In [19]:
# Ex III.3
space_df.loc[[1318, 2976, 4131], ["Datum", "Company Name"]]

Datum Company Name
1318  Tue Apr 28, 1998 22:53 UTC  Arianespace
2976  Sat Dec 21, 1974 02:19 UTC    RVSN USSR
4131  Sat Sep 01, 1962 02:12 UTC    RVSN USSR

#### Ex III.4
Display the final 10 rows of the *Detail* column.

In [20]:
# Ex III.4
space_df.loc[-10:, "Detail"]

0       Falcon 9 Block 5 | Starlink V1 L9 & BlackSky
1                Long March 2D | Gaofen-9 04 & Q-SAT
2                 Starship Prototype | 150 Meter Hop
3       Proton-M/Briz-M | Ekspress-80 & Ekspress-103
4                         Atlas V 541 | Perseverance
                            ...                     
4319                       Vanguard | Vanguard TV3BU
4320                             Juno I | Explorer 1
4321                         Vanguard | Vanguard TV3
4322                      Sputnik 8K71PS | Sputnik-2
4323                      Sputnik 8K71PS | Sputnik-1
Name: Detail, Length: 4324, dtype: object

### F. Searching Within a Dataframe
Being able to extract data by row and column numbers is helpful at times, but it requires that we know the exact location of the data we want. In large dataframes with thousands of rows, we typically do NOT know the exact location. Fortunately, Pandas provides many techniques for searching within a dataframe

#### 1. Boolean Indexing
Boolean indexing is a powerful technique for filtering dataframes. Check out the following example.

In [21]:
# Searching for Specific Data in Dataframe
# How many successful space launches were conducted by the U.S. Navy?
space_df[space_df["Company Name"] == "US Navy"].head()

Company Name                                  Location  \
4277      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   
4285      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   
4288      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   
4291      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   
4300      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   

                           Datum                     Detail  Status Rocket  \
4277  Fri Sep 18, 1959 05:20 UTC      Vanguard | Vanguard 3  StatusRetired   
4285  Mon Jun 22, 1959 20:16 UTC  Vanguard | Vanguard SLV-6  StatusRetired   
4288  Tue Apr 14, 1959 02:49 UTC  Vanguard | Vanguard SLV-5  StatusRetired   
4291  Tue Feb 17, 1959 15:55 UTC      Vanguard | Vanguard 2  StatusRetired   
4300  Fri Sep 26, 1958 15:38 UTC  Vanguard | Vanguard SLV-3  StatusRetired   

      Rocket   Status Mission  
4277     NaN  Partial Failure  
4285     NaN          Failure  
4288     NaN          Failure  
4291     NaN          Success  
4300     NaN          Failure

The syntax in for filtering the dataframe to only US Navy launches may look strange. It will make sense once we know more about how Pandas works. First, let's see what the expression `space_df["Company Name"] == "US Navy"` does by itself.

In [22]:
print('Expression: space_df["Company Name"] == "US Navy"')
print("Type of object returned by expression:", type(space_df["Company Name"] == "US Navy"))
print("Object contents:")
space_df["Company Name"] == "US Navy"

Expression: space_df["Company Name"] == "US Navy"
Type of object returned by expression: <class 'pandas.core.series.Series'>
Object contents:


0       False
1       False
2       False
3       False
4       False
        ...  
4319     True
4320    False
4321     True
4322    False
4323    False
Name: Company Name, Length: 4324, dtype: bool

The expression returned `Series` object of Boolean values. The Series has a length of 4,324 items, which is the same as the number of rows in the dataframe. That's suspicous...

In the expression, we're taking the `Series` object returned by the expression `space_df["Company Name"]` and we're comparing it to the string "US Navy". Pandas then checks every single element (all 4,324) to the string "US Navy", generating a value of True if the value is equal to "US Navy" and False otherwise. This action generates 4,234 True or False values, which Pandas returns as a `Series`. An operation that is repeated on every single element of an array-like object (e.g., list, array, Series) is called an element-wise or vectorized operation. One of the benefits of using Pandas is it is capable of many element-wise operations. Not having to write a `for` loop is convenient and makes code more concise. Also, element-wise operations are often faster than operations that use `for` loops.

When we pass a `Series` of Boolean values to the dataframe by placing it in square brackets, Pandas will return another dataframe with only the rows that correspond to values of True in the Boolean series. Let's experiment. The following cell generates a Boolean series that is the same length as `space_df`, with every value set to `False`.

In [23]:
# Create a series where every element is equal to False
bseries = pd.Series([False] * len(space_df))
bseries

0       False
1       False
2       False
3       False
4       False
        ...  
4319    False
4320    False
4321    False
4322    False
4323    False
Length: 4324, dtype: bool

Now we'll bass the series to the space_df dataframe.

In [24]:
# Pass series to space_df
space_df[bseries]

Empty DataFrame
Columns: [Company Name, Location, Datum, Detail, Status Rocket, Rocket, Status Mission]
Index: []

As expected, no dataframe rows were returned, because all `bseries` elements are false. Let's change a few of the elements to `True`.

In [25]:
bseries[1318] = True
bseries[1595] = True
bseries[2046] = True
bseries[2907] = True
bseries[2017] = True
space_df[bseries]

Company Name                                           Location  \
1318  Arianespace  ELA-2, Guiana Space Centre, French Guiana, France   
1595  Arianespace  ELA-2, Guiana Space Centre, French Guiana, France   
2017    RVSN USSR            Site 132/2, Plesetsk Cosmodrome, Russia   
2046    RVSN USSR             Site 43/4, Plesetsk Cosmodrome, Russia   
2907         CASC  Site 138 (LA-2B), Jiuquan Satellite Launch Cen...   

                           Datum                                  Detail  \
1318  Tue Apr 28, 1998 22:53 UTC       Ariane 44P | Nilesat-101, BSAT-1B   
1595  Sat Dec 18, 1993 01:27 UTC           Ariane 44L | DBS-1, Thaicom-1   
2017  Thu Jan 29, 1987 06:14 UTC        Cosmos-3M (11K65M) | Cosmos 1816   
2046  Fri Sep 05, 1986 09:12 UTC  Molniya-M /Block ML | Molniya-1 n†­125   
2907  Sat Jul 26, 1975 13:30 UTC                     Feng Bao 1 | JSSW-3   

      Status Rocket  Rocket   Status Mission  
1318  StatusRetired     NaN          Success  
1595  StatusRetired     NaN          Success  
2017  StatusRetired     NaN  Partial Failure  
2046  StatusRetired     NaN          Success  
2907  StatusRetired     NaN          Success

For comparison operators we are not limited to `==`. We can also use `<=`, `>=`, `<`, `>`, and `!=`.

#### 2. Compound Boolean Indexing
Compound Boolean exressions are also allowed. See the following example.

In [26]:
# Searching for Specific Data in Dataframe
# How many unsuccessful space launches were conducted by the U.S. Navy?
space_df[(space_df["Company Name"] == "US Navy") & (space_df["Status Mission"] == "Success")]

Company Name                                  Location  \
4291      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   
4317      US Navy  LC-18A, Cape Canaveral AFS, Florida, USA   

                           Datum                 Detail  Status Rocket  \
4291  Tue Feb 17, 1959 15:55 UTC  Vanguard | Vanguard 2  StatusRetired   
4317  Mon Mar 17, 1958 12:15 UTC  Vanguard | Vanguard 1  StatusRetired   

      Rocket Status Mission  
4291     NaN        Success  
4317     NaN        Success

Please note that the Boolean indexing syntax uses different Boolean operators than standard Python code. Instead of `and`, Boolean indexing uses `&`. Instead of `or` and `not`, Boolean indexing uses `|` and `~` respectively. Also, each part of a compound Boolean expression *must* be grouped with parentheses, or the expression will generate an error.

The `isin()` method is useful.

In [27]:
space_df[space_df["Company Name"].isin(["SpaceX", "Blue Origin", "ULA"])]

Company Name                                           Location  \
0         SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
2         SpaceX                      Pad A, Boca Chica, Texas, USA   
4            ULA           SLC-41, Cape Canaveral AFS, Florida, USA   
8         SpaceX           SLC-40, Cape Canaveral AFS, Florida, USA   
17        SpaceX           SLC-40, Cape Canaveral AFS, Florida, USA   
..           ...                                                ...   
879       SpaceX  Omelek Island, Ronald Reagan Ballistic Missile...   
881          ULA           SLC-41, Cape Canaveral AFS, Florida, USA   
883          ULA          SLC-17B, Cape Canaveral AFS, Florida, USA   
890          ULA            SLC-2W, Vandenberg AFB, California, USA   
929       SpaceX  Omelek Island, Ronald Reagan Ballistic Missile...   

                          Datum                                        Detail  \
0    Fri Aug 07, 2020 05:12 UTC  Falcon 9 Block 5 | Starlink V1 L9 & BlackSky   
2    Tue Aug 04, 2020 23:57 UTC            Starship Prototype | 150 Meter Hop   
4    Thu Jul 30, 2020 11:50 UTC                    Atlas V 541 | Perseverance   
8    Mon Jul 20, 2020 21:30 UTC                  Falcon 9 Block 5 | ANASIS-II   
17   Tue Jun 30, 2020 20:10 UTC               Falcon 9 Block 5 | GPS III SV03   
..                          ...                                           ...   
879  Wed Mar 21, 2007 01:10 UTC                            Falcon 1 | DemoSat   
881  Fri Mar 09, 2007 03:10 UTC              Atlas V 401 | STP-1, FalconSat-3   
883  Sat Feb 17, 2007 23:01 UTC                    Delta II 7925-10C | THEMIS   
890  Thu Dec 14, 2006 21:00 UTC                   Delta II 7920-10C | NROL-21   
929  Fri Mar 24, 2006 21:30 UTC                        Falcon 1 | FalconSat-2   

     Status Rocket  Rocket Status Mission  
0     StatusActive    50.0        Success  
2     StatusActive     NaN        Success  
4     StatusActive   145.0        Success  
8     StatusActive    50.0        Success  
17    StatusActive    50.0        Success  
..             ...     ...            ...  
879  StatusRetired     7.0        Failure  
881   StatusActive   109.0        Success  
883  StatusRetired     NaN        Success  
890  StatusRetired     NaN        Success  
929  StatusRetired     7.0        Failure  

[252 rows x 7 columns]

#### 3. The `isin()` Method
The `.isin()` method returns `True` if the value is equal to any of the elements in the list that is passed as a parameter. We could get the same results with a long compound Boolean statement with many `|` (i.e., *or*) operators, but using `.isin()` is much easier. The result can be negated with the `~` operator.

In [28]:
space_df[~space_df["Company Name"].isin(["SpaceX", "Blue Origin", "ULA"])]

Company Name                                           Location  \
1            CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
3       Roscosmos       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
5            CASC       LC-9, Taiyuan Satellite Launch Center, China   
6       Roscosmos         Site 31/6, Baikonur Cosmodrome, Kazakhstan   
7            CASC    LC-101, Wenchang Satellite Launch Center, China   
...           ...                                                ...   
4319      US Navy           LC-18A, Cape Canaveral AFS, Florida, USA   
4320         AMBA           LC-26A, Cape Canaveral AFS, Florida, USA   
4321      US Navy           LC-18A, Cape Canaveral AFS, Florida, USA   
4322    RVSN USSR          Site 1/5, Baikonur Cosmodrome, Kazakhstan   
4323    RVSN USSR          Site 1/5, Baikonur Cosmodrome, Kazakhstan   

                           Datum  \
1     Thu Aug 06, 2020 04:01 UTC   
3     Thu Jul 30, 2020 21:25 UTC   
5     Sat Jul 25, 2020 03:13 UTC   
6     Thu Jul 23, 2020 14:26 UTC   
7     Thu Jul 23, 2020 04:41 UTC   
...                          ...   
4319  Wed Feb 05, 1958 07:33 UTC   
4320  Sat Feb 01, 1958 03:48 UTC   
4321  Fri Dec 06, 1957 16:44 UTC   
4322  Sun Nov 03, 1957 02:30 UTC   
4323  Fri Oct 04, 1957 19:28 UTC   

                                                 Detail  Status Rocket  \
1                   Long March 2D | Gaofen-9 04 & Q-SAT   StatusActive   
3          Proton-M/Briz-M | Ekspress-80 & Ekspress-103   StatusActive   
5     Long March 4B | Ziyuan-3 03, Apocalypse-10 & N...   StatusActive   
6                           Soyuz 2.1a | Progress MS-15   StatusActive   
7                              Long March 5 | Tianwen-1   StatusActive   
...                                                 ...            ...   
4319                          Vanguard | Vanguard TV3BU  StatusRetired   
4320                                Juno I | Explorer 1  StatusRetired   
4321                            Vanguard | Vanguard TV3  StatusRetired   
4322                         Sputnik 8K71PS | Sputnik-2  StatusRetired   
4323                         Sputnik 8K71PS | Sputnik-1  StatusRetired   

      Rocket Status Mission  
1      29.75        Success  
3      65.00        Success  
5      64.68        Success  
6      48.50        Success  
7        NaN        Success  
...      ...            ...  
4319     NaN        Failure  
4320     NaN        Success  
4321     NaN        Failure  
4322     NaN        Success  
4323     NaN        Success  

[4072 rows x 7 columns]

##### More Information
* [Boolean Indexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#boolean-indexing)

#### 4. *NaN* Values
By now you've noticed that the value *Nan* appears frequently. It stands for *not a number* and is used to represent missing data. We can use the `.notna()` method to filter out rows with *NaN* values. Suppose we only want rows with no missing data in the *Rocket* column:

In [29]:
space_df[space_df["Rocket"].notna()].head()

Company Name                                           Location  \
0       SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
1         CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
3    Roscosmos       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4          ULA           SLC-41, Cape Canaveral AFS, Florida, USA   
5         CASC       LC-9, Taiyuan Satellite Launch Center, China   

                        Datum  \
0  Fri Aug 07, 2020 05:12 UTC   
1  Thu Aug 06, 2020 04:01 UTC   
3  Thu Jul 30, 2020 21:25 UTC   
4  Thu Jul 30, 2020 11:50 UTC   
5  Sat Jul 25, 2020 03:13 UTC   

                                              Detail Status Rocket  Rocket  \
0       Falcon 9 Block 5 | Starlink V1 L9 & BlackSky  StatusActive   50.00   
1                Long March 2D | Gaofen-9 04 & Q-SAT  StatusActive   29.75   
3       Proton-M/Briz-M | Ekspress-80 & Ekspress-103  StatusActive   65.00   
4                         Atlas V 541 | Perseverance  StatusActive  145.00   
5  Long March 4B | Ziyuan-3 03, Apocalypse-10 & N...  StatusActive   64.68   

  Status Mission  
0        Success  
1        Success  
3        Success  
4        Success  
5        Success

Conversely, we can use the `.isna()` method to select rows that contain *NaN*.

##### More Information
* [Pandas documentation on missing data](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html)

#### 5. Using the Query Method

The `.query()` method can also be used to extract specific rows from a dataframe. Users pass a query string to the method instead of a Boolean series object. Our queries will be more effective if we make a couple tweaks on the space dataframe first. All dataframe columns are currently strings. The following cell will convert two of the columns to different data types.

In [31]:
# Convert Rocket and Datum columns to numeric and datetime data types
space_df["Rocket"] = space_df["Rocket"].astype('float32')
space_df["Datum"] = pd.to_datetime(space_df["Datum"])

ValueError: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True

The next cell uses the `.query()` method to find all launches by NASA where the mission cost more than $1,000,000,000 (The *Rocket* column contains the cost of the mission, in millions of dollars).

In [62]:
# Alternate search technique using the `query()` method
space_df.query("`Company Name` == 'NASA' and Rocket > 1000")

Company Name                                    Location  \
3149         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3180         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3243         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3328         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3384         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3475         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3511         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3545         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3560         NASA  LC-39B, Kennedy Space Center, Florida, USA   
3584         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3603         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3683         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3716         NASA  LC-39A, Kennedy Space Center, Florida, USA   

                          Datum                Detail  Status Rocket  Rocket  \
3149  1973-05-14 17:30:00+00:00   Saturn V | Skylab 1  StatusRetired  1160.0   
3180  1972-12-19 19:24:00+00:00  Saturn V | Apollo 17  StatusRetired  1160.0   
3243  1972-04-16 17:54:00+00:00  Saturn V | Apollo 16  StatusRetired  1160.0   
3328  1971-07-26 13:34:00+00:00  Saturn V | Apollo 15  StatusRetired  1160.0   
3384  1971-01-31 21:03:00+00:00  Saturn V | Apollo 14  StatusRetired  1160.0   
3475  1970-04-11 19:13:00+00:00  Saturn V | Apollo 13  StatusRetired  1160.0   
3511  1969-11-14 16:22:00+00:00  Saturn V | Apollo 12  StatusRetired  1160.0   
3545  1969-07-16 13:32:00+00:00  Saturn V | Apollo 11  StatusRetired  1160.0   
3560  1969-05-18 16:49:00+00:00  Saturn V | Apollo 10  StatusRetired  1160.0   
3584  1969-03-03 16:00:00+00:00   Saturn V | Apollo 9  StatusRetired  1160.0   
3603  1968-12-21 12:51:00+00:00   Saturn V | Apollo 8  StatusRetired  1160.0   
3683  1968-04-04 12:00:00+00:00   Saturn V | Apollo 6  StatusRetired  1160.0   
3716  1967-11-09 12:00:00+00:00   Saturn V | Apollo 4  StatusRetired  1160.0   

       Status Mission  
3149          Success  
3180          Success  
3243          Success  
3328          Success  
3384          Success  
3475          Success  
3511          Success  
3545          Success  
3560          Success  
3584          Success  
3603          Success  
3683  Partial Failure  
3716          Success

Query strings are more intuitive and less verbose than Boolean indexing. They can also be faster on large dataframes. They do have a few quirks:
* The query string specifies criteria that each column name must meet in order to be included in the resulting dataframe. If column names that contain spaces or other special characters must be enclosed in backticks, like `Company Name`. Note that a backtick (\`) is not the same as a single quote ('). The backtick key on your keyboard should be near th upper left corner, whereas the single quote is next to the ENTER key.
* String values that appear in the query, like 'NASA', must be quoted. We use single quotes to denote strings in our example because the entire query string is enclosed in double quotes. But we could swap the single and double quotes --  '`Company Name` == "NASA" and Rocket > 1000' would work just as well as a query string.
* Query strings uses `and`, `or`, and `not` as logical operators.
* Comparison operators in query strings are as expected: `==`, `<=`, `>=`, `<`, `>`, and `!=`.

Suppose the query was embedded in a function, with the *Company Name* passed in as a parameter. We can include a variable in the query string using the 

In [33]:
def get_launches(company):
    return space_df.query("`Company Name` == @company and Rocket > 1000")

get_launches("NASA").head()

Company Name                                    Location  \
3149         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3180         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3243         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3328         NASA  LC-39A, Kennedy Space Center, Florida, USA   
3384         NASA  LC-39A, Kennedy Space Center, Florida, USA   

                          Datum                Detail  Status Rocket  Rocket  \
3149  1973-05-14 17:30:00+00:00   Saturn V | Skylab 1  StatusRetired  1160.0   
3180  1972-12-19 19:24:00+00:00  Saturn V | Apollo 17  StatusRetired  1160.0   
3243  1972-04-16 17:54:00+00:00  Saturn V | Apollo 16  StatusRetired  1160.0   
3328  1971-07-26 13:34:00+00:00  Saturn V | Apollo 15  StatusRetired  1160.0   
3384  1971-01-31 21:03:00+00:00  Saturn V | Apollo 14  StatusRetired  1160.0   

     Status Mission  
3149        Success  
3180        Success  
3243        Success  
3328        Success  
3384        Success

The `.query()` method accepts `in` and `not in` operators, similar to Boolean indexing.

In [34]:
companies = ["SpaceX", "Blue Origin", "ULA"]
space_df.query("`Company Name` in @companies").head()

Company Name                                    Location  \
0        SpaceX  LC-39A, Kennedy Space Center, Florida, USA   
2        SpaceX               Pad A, Boca Chica, Texas, USA   
4           ULA    SLC-41, Cape Canaveral AFS, Florida, USA   
8        SpaceX    SLC-40, Cape Canaveral AFS, Florida, USA   
17       SpaceX    SLC-40, Cape Canaveral AFS, Florida, USA   

                        Datum                                        Detail  \
0   2020-08-07 05:12:00+00:00  Falcon 9 Block 5 | Starlink V1 L9 & BlackSky   
2   2020-08-04 23:57:00+00:00            Starship Prototype | 150 Meter Hop   
4   2020-07-30 11:50:00+00:00                    Atlas V 541 | Perseverance   
8   2020-07-20 21:30:00+00:00                  Falcon 9 Block 5 | ANASIS-II   
17  2020-06-30 20:10:00+00:00               Falcon 9 Block 5 | GPS III SV03   

   Status Rocket  Rocket Status Mission  
0   StatusActive    50.0        Success  
2   StatusActive     NaN        Success  
4   StatusActive   145.0        Success  
8   StatusActive    50.0        Success  
17  StatusActive    50.0        Success

##### More Information
* [Pandas user guide section on the `.query()` method](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#the-query-method).
* [Query method API reference](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html#pandas.DataFrame.query)

### 6. Boolean Indexing Exercises and the Query Method

##### Ex III.5
The `grads` dataframe contains information on different college majors. How many majors are there in the *Engineering* category? How many in *Biology and Life Science*?

In [35]:
# Ex III.5
grads = pd.read_csv("recent-grads.csv")     # Leave this line alone
grads[grads["Major_category"] == "Engineering"]


Rank  Major_code                                        Major    Total  \
0      1        2419                        PETROLEUM ENGINEERING   2339.0   
1      2        2416               MINING AND MINERAL ENGINEERING    756.0   
2      3        2415                    METALLURGICAL ENGINEERING    856.0   
3      4        2417    NAVAL ARCHITECTURE AND MARINE ENGINEERING   1258.0   
4      5        2405                         CHEMICAL ENGINEERING  32260.0   
5      6        2418                          NUCLEAR ENGINEERING   2573.0   
8      9        2414                       MECHANICAL ENGINEERING  91227.0   
9     10        2408                       ELECTRICAL ENGINEERING  81527.0   
10    11        2407                         COMPUTER ENGINEERING  41542.0   
11    12        2401                        AEROSPACE ENGINEERING  15058.0   
12    13        2404                       BIOMEDICAL ENGINEERING  14955.0   
13    14        5008                            MATERIALS SCIENCE   4279.0   
14    15        2409    ENGINEERING MECHANICS PHYSICS AND SCIENCE   4321.0   
15    16        2402                       BIOLOGICAL ENGINEERING   8925.0   
16    17        2412     INDUSTRIAL AND MANUFACTURING ENGINEERING  18968.0   
17    18        2400                          GENERAL ENGINEERING  61152.0   
18    19        2403                    ARCHITECTURAL ENGINEERING   2825.0   
22    23        2502            ELECTRICAL ENGINEERING TECHNOLOGY  11565.0   
23    24        2413  MATERIALS ENGINEERING AND MATERIALS SCIENCE   2993.0   
25    26        2406                            CIVIL ENGINEERING  53153.0   
28    29        2499                    MISCELLANEOUS ENGINEERING   9133.0   
30    31        2410                    ENVIRONMENTAL ENGINEERING   4047.0   
31    32        2500                     ENGINEERING TECHNOLOGIES   3600.0   
33    34        2411       GEOLOGICAL AND GEOPHYSICAL ENGINEERING    720.0   
38    39        2503           INDUSTRIAL PRODUCTION TECHNOLOGIES   4631.0   
50    51        2501        ENGINEERING AND INDUSTRIAL MANAGEMENT   2906.0   
58    59        1401                                 ARCHITECTURE  46420.0   
65    66        2599       MISCELLANEOUS ENGINEERING TECHNOLOGIES   8804.0   
66    67        2504  MECHANICAL ENGINEERING RELATED TECHNOLOGIES   4790.0   

        Men    Women Major_category  ShareWomen  Sample_size  Employed  ...  \
0    2057.0    282.0    Engineering    0.120564           36      1976  ...   
1     679.0     77.0    Engineering    0.101852            7       640  ...   
2     725.0    131.0    Engineering    0.153037            3       648  ...   
3    1123.0    135.0    Engineering    0.107313           16       758  ...   
4   21239.0  11021.0    Engineering    0.341631          289     25694  ...   
5    2200.0    373.0    Engineering    0.144967           17      1857  ...   
8   80320.0  10907.0    Engineering    0.119559         1029     76442  ...   
9   65511.0  16016.0    Engineering    0.196450          631     61928  ...   
10  33258.0   8284.0    Engineering    0.199413          399     32506  ...   
11  12953.0   2105.0    Engineering    0.139793          147     11391  ...   
12   8407.0   6548.0    Engineering    0.437847           79     10047  ...   
13   2949.0   1330.0    Engineering    0.310820           22      3307  ...   
14   3526.0    795.0    Engineering    0.183985           30      3608  ...   
15   6062.0   2863.0    Engineering    0.320784           55      6170  ...   
16  12453.0   6515.0    Engineering    0.343473          183     15604  ...   
17  45683.0  15469.0    Engineering    0.252960          425     44931  ...   
18   1835.0    990.0    Engineering    0.350442           26      2575  ...   
22   8181.0   3384.0    Engineering    0.292607           97      8587  ...   
23   2020.0    973.0    Engineering    0.325092           22      2449  ...   
25  41081.0  12072.0    Engineering    0.227118          565     43041  ...   
28  

##### Ex III.6
Display the first 20 rows and only the *Major*, *Total*, *Unemployment_rate*, and *Median* columns of the grads dataframe.

In [36]:
# Ex III.6
grads[["Major", "Total", "Unemployment_rate", "Median"]].head(20)


Major    Total  Unemployment_rate  \
0                       PETROLEUM ENGINEERING   2339.0           0.018381   
1              MINING AND MINERAL ENGINEERING    756.0           0.117241   
2                   METALLURGICAL ENGINEERING    856.0           0.024096   
3   NAVAL ARCHITECTURE AND MARINE ENGINEERING   1258.0           0.050125   
4                        CHEMICAL ENGINEERING  32260.0           0.061098   
5                         NUCLEAR ENGINEERING   2573.0           0.177226   
6                           ACTUARIAL SCIENCE   3777.0           0.095652   
7                  ASTRONOMY AND ASTROPHYSICS   1792.0           0.021167   
8                      MECHANICAL ENGINEERING  91227.0           0.057342   
9                      ELECTRICAL ENGINEERING  81527.0           0.059174   
10                       COMPUTER ENGINEERING  41542.0           0.065409   
11                      AEROSPACE ENGINEERING  15058.0           0.065162   
12                     BIOMEDICAL ENGINEERING  14955.0           0.092084   
13                          MATERIALS SCIENCE   4279.0           0.023043   
14  ENGINEERING MECHANICS PHYSICS AND SCIENCE   4321.0           0.006334   
15                     BIOLOGICAL ENGINEERING   8925.0           0.087143   
16   INDUSTRIAL AND MANUFACTURING ENGINEERING  18968.0           0.042876   
17                        GENERAL ENGINEERING  61152.0           0.059824   
18                  ARCHITECTURAL ENGINEERING   2825.0           0.061931   
19                            COURT REPORTING   1148.0           0.011690   

    Median  
0   110000  
1    75000  
2    73000  
3    70000  
4    65000  
5    65000  
6    62000  
7    62000  
8    60000  
9    60000  
10   60000  
11   60000  
12   60000  
13   60000  
14   58000  
15   57100  
16   57000  
17   56000  
18   54000  
19   54000

##### Ex III.7

Use the `.sort_values()` method to sort the `grads` dataframe by *Median* in descending order and display the first five rows. Which major has the highest median income? Refer to the [Dataframe.sort_values documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html?highlight=sort_values#pandas.DataFrame.sort_values).

In [38]:
# Ex III.7
grads["Median"].sort_values(ascending = False).tail(5)


168    26000
169    25000
170    25000
171    23400
172    22000
Name: Median, dtype: int64

##### Ex III.8
Use Boolean indexing to extract majors in the categories of *Business*, *Physical Sciences*, or *Arts* that have greater than 2500 respondents (*Total* column). Don't forget the parentheses.

In [39]:
# Ex III.8
grads.query("Major_category in ['Business', 'Physical Sciences', 'Arts'] and 'Major_code' > '2500'")

Rank  Major_code                                              Major  \
6       7        6202                                  ACTUARIAL SCIENCE   
7       8        5001                         ASTRONOMY AND ASTROPHYSICS   
24     25        6212      MANAGEMENT INFORMATION SYSTEMS AND STATISTICS   
27     28        6204                OPERATIONS LOGISTICS AND E-COMMERCE   
32     33        6099                            MISCELLANEOUS FINE ARTS   
35     36        6207                                            FINANCE   
37     38        6205                                 BUSINESS ECONOMICS   
39     40        5102  NUCLEAR, INDUSTRIAL RADIOLOGY, AND BIOLOGICAL ...   
40     41        6201                                         ACCOUNTING   
43     44        5007                                            PHYSICS   
49     50        5006                                       OCEANOGRAPHY   
57     58        6200                                   GENERAL BUSINESS   
59     60        6210                             INTERNATIONAL BUSINESS   
62     63        6299    MISCELLANEOUS BUSINESS & MEDICAL ADMINISTRATION   
72     73        5000                                  PHYSICAL SCIENCES   
74     75        5003                                          CHEMISTRY   
76     77        6203             BUSINESS MANAGEMENT AND ADMINISTRATION   
77     78        6206                   MARKETING AND MARKETING RESEARCH   
85     86        5004                          GEOLOGY AND EARTH SCIENCE   
86     87        6209           HUMAN RESOURCES AND PERSONNEL MANAGEMENT   
90     91        5005                                        GEOSCIENCES   
95     96        6004                  COMMERCIAL ART AND GRAPHIC DESIGN   
97     98        5098              MULTI-DISCIPLINARY OR GENERAL SCIENCE   
110   111        5002               ATMOSPHERIC SCIENCES AND METEOROLOGY   
127   128        6211                             HOSPITALITY MANAGEMENT   
141   142        6005                   FILM VIDEO AND PHOTOGRAPHIC ARTS   
146   147        6002                                              MUSIC   
149   150        6000                                          FINE ARTS   
153   154        6003                         VISUAL AND PERFORMING ARTS   
159   160        6007                                        STUDIO ARTS   
166   167        6001                             DRAMA AND THEATER ARTS   

        Total       Men     Women     Major_category  ShareWomen  Sample_size  \
6      3777.0    2110.0    1667.0           Business    0.441356           51   
7      1792.0     832.0     960.0  Physical Sciences    0.535714           10   
24    18713.0   13496.0    5217.0           Business    0.278790          278   
27    11732.0    7921.0    3811.0           Business    0.324838          156   
32     3340.0    1970.0    1370.0               Arts    0.410180           30   
35   174506.0  115030.0   59476.0           Business    0.340825         2189   
37    13302.0    7575.0    5727.0           Business    0.430537          199   
39     2116.0     528.0    1588.0  Physical Sciences    0.750473           31   
40   198633.0   94519.0  104114.0           Business    0.524153         2042   
43    32142.0   23080.0    9062.0  Physical Sciences    0.281936          142   
49     2418.0     752.0    1666.0  Physical Sciences    0.688999           36   
57   234590.0  132238.0  102352.0           Business    0.436302         2380   
59    25894.0   10624.0   15270.0           Business    0.589712          260   
62    17947.0   10285.0    7662.0           Business    0.426924          244   
72     1436.0     894.0     542.0  Physical Sciences    0.377437           10   
74    66530.0   32923.0   33607.0  Physical Sciences    0.505141          353   
76   329927.0  173809.0  156118.0           Business    0.473190         4212   
77   205211.0   78857.0  126354.0           Business    0.615727         2684   
85    10972.0    5813.0    5159.0

##### Ex III.9
Use the `.query()` method to identify majors for which the share of women is greater than 50% and have a median salary greater $50,000.

In [40]:
# Ex III.9
grads.query("ShareWomen > 0.5 and Median > 50000")

Rank  Major_code                       Major   Total    Men  Women  \
7     8        5001  ASTRONOMY AND ASTROPHYSICS  1792.0  832.0  960.0   

      Major_category  ShareWomen  Sample_size  Employed  ...  Part_time  \
7  Physical Sciences    0.535714           10      1526  ...        553   

   Full_time_year_round  Unemployed  Unemployment_rate  Median  P25th   P75th  \
7                   827          33           0.021167   62000  31500  109000   

   College_jobs  Non_college_jobs  Low_wage_jobs  
7           972               500            220  

[1 rows x 21 columns]

### E. Column Datatypes
Every column in a Pandas `DataFrame` has a specific datatype. For the sake of demonstration, we will reload the space dataframe from the CSV file.

In [41]:
# Reload space dataframe
space_df2 = pd.read_csv("space.csv")

We can see the datatype of each column with the `.dtypes` attribute.

In [42]:
print("Column Datatypes")
space_df2.dtypes

Column Datatypes


Company Name      object
Location          object
Datum             object
Detail            object
Status Rocket     object
Rocket            object
Status Mission    object
dtype: object

The type *object* is Pandas' datatype for a string. You might think that the *Rocket* column should be a numeric datatype. Pandas decided it was a string because several entries are using a comma as a thousands separator, e.g., `1,160`. The `.read_csv` method has an argument that will help Pandas read the *Rocket* column as a string, in spite of the commas.

In [43]:
# Reload space dataframe, with commas as thousands separator
space_df2 = pd.read_csv("space.csv", thousands=",")

print("Column Datatypes")
space_df2.dtypes

Column Datatypes


Company Name       object
Location           object
Datum              object
Detail             object
Status Rocket      object
Rocket            float64
Status Mission     object
dtype: object

The *Rocket* column is now a floating point number. That's more like it.

But what about the Datum column? We can't do date calculation on that column because Pandas is considering it to be a string. We can convert the column to a special datetime object with the `pandas.to_datetime()` method.

In [44]:
# Convert Datum to a datetime object.
space_df2["Datum"] = pd.to_datetime(space_df2["Datum"], utc=True)
print("Column Datatypes")
space_df2.dtypes

Column Datatypes


Company Name                   object
Location                       object
Datum             datetime64[ns, UTC]
Detail                         object
Status Rocket                  object
Rocket                        float64
Status Mission                 object
dtype: object

Now that the column is converted to a datetime object, we can access parts of the date using the `.dt` accessor. For example, to get an integer representing the month:

In [45]:
space_df2.Datum.dt.month.head()

0    8
1    8
2    8
3    7
4    7
Name: Datum, dtype: int64

Another useful method is `.astype()`. For example:

In [46]:
space_df2.Rocket.astype("float32")

0        50.00
1        29.75
2          NaN
3        65.00
4       145.00
         ...  
4319       NaN
4320       NaN
4321       NaN
4322       NaN
4323       NaN
Name: Rocket, Length: 4324, dtype: float32

We converted the numbers in the *Rocket* column to 32-bit numbers (from 64-bit numbers) to save space. Other common datatypes include "int32", "int64", "unint32" (unsigned, meaning only positve numbers are allowed), and "boolean".

##### More Information
* [Pandas Datatypes](https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html#dtypes)

#### 1. Datatype Exercises

##### Ex III.10
Using the `grads` dataframe, convert the *Median* column's datatype to `uint32`.

In [47]:
# Ex III.10
grads.Median.astype("uint32")

0      110000
1       75000
2       73000
3       70000
4       65000
        ...  
168     26000
169     25000
170     25000
171     23400
172     22000
Name: Median, Length: 173, dtype: uint32

### E. Statistics Functions
Pandas has several functions for extracting summary statistics.

In [48]:
print("Maximum mission cost in $millions:\t", space_df.Rocket.max())
print("Mean mission cost in $millions:\t\t", space_df.Rocket.mean())
print("Standard Deviation mission cost:\t", space_df.Rocket.std())

Maximum mission cost in $millions:	 5000.0
Mean mission cost in $millions:		 153.79217529296875
Standard Deviation mission cost:	 288.45074462890625


##### More Information
* [Getting Started Tutorial on Summary Satistics](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/06_calculate_statistics.html)

#### 1. Statistics Exercises

##### Ex III.11
Use the `.min()` and `max()` methods to determine which majors have the lowest and highest unemployment rates.

In [49]:
# Ex III.11
min = grads.Unemployment_rate.min()
max = grads.Unemployment_rate.max()
grads.query("`Unemployment_rate` == @min or `Unemployment_rate` == @max")

Rank  Major_code                                       Major   Total  \
5       6        2418                         NUCLEAR ENGINEERING  2573.0   
52     53        4005            MATHEMATICS AND COMPUTER SCIENCE   609.0   
73     74        3801                       MILITARY TECHNOLOGIES   124.0   
83     84        3602                                      BOTANY  1329.0   
112   113        1106                                SOIL SCIENCE   685.0   
120   121        2301  EDUCATIONAL ADMINISTRATION AND SUPERVISION   804.0   

        Men  Women                       Major_category  ShareWomen  \
5    2200.0  373.0                          Engineering    0.144967   
52    500.0  109.0              Computers & Mathematics    0.178982   
73    124.0    0.0  Industrial Arts & Consumer Services    0.000000   
83    626.0  703.0               Biology & Life Science    0.528969   
112   476.0  209.0      Agriculture & Natural Resources    0.305109   
120   280.0  524.0                            Education    0.651741   

     Sample_size  Employed  ...  Part_time  Full_time_year_round  Unemployed  \
5             17      1857  ...        264                  1449         400   
52             7       559  ...          0                   391           0   
73             4         0  ...          0                   111           0   
83             9      1010  ...        169                   740           0   
112            4       613  ...        185                   383           0   
120            5       703  ...          0                   504           0   

     Unemployment_rate  Median  P25th   P75th  College_jobs  Non_college_jobs  \
5             0.177226   65000  50000  102000          1142               657   
52            0.000000   42000  30000   78000           452                67   
73            0.000000   40000  40000   40000             0                 0   
83            0.000000   37000  26000   40000           677               184   
112           0.000000   35000  18500   44000           355               144   
120           0.000000   34000  29000   35000           346               206   

     Low_wage_jobs  
5              244  
52              25  
73               0  
83              56  
112              0  
120            111  

[6 rows x 21 columns]

##### Ex III.12
Repeat exercise III.11, but only consider majors with more than 2,000 respondents (*Total* column).

In [50]:
# Ex III.12
grads.query("(`Unemployment_rate` == @min or `Unemployment_rate` == @max) and Total >= 2000")

Rank  Major_code                Major   Total     Men  Women  \
5     6        2418  NUCLEAR ENGINEERING  2573.0  2200.0  373.0   

  Major_category  ShareWomen  Sample_size  Employed  ...  Part_time  \
5    Engineering    0.144967           17      1857  ...        264   

   Full_time_year_round  Unemployed  Unemployment_rate  Median  P25th   P75th  \
5                  1449         400           0.177226   65000  50000  102000   

   College_jobs  Non_college_jobs  Low_wage_jobs  
5          1142               657            244  

[1 rows x 21 columns]

### G. Dataframe Indexes

#### 1. Getting started with Indexes
Indexes are an important concept in Pandas. To understand indexes, we'll use a dataframe that contains match data from the 2019 Pacific Northwest district competition at Glacier Peak High School in Snohomish, WA. First we have to load the dataframe from a file.

In [89]:
# Loading a datframe from a pickle file.
import pickle
with open("matches.pickle", "rb") as pfile:
    matches = pickle.load(pfile)

The Python Standard Library includes a *pickle* module that allows us to save any Python object to a file and load it back into Python at a later time. Here are the first few rows from the dataframe that was contained in the *matches.pickle* file.

In [55]:
matches.head(3)

actual_time comp_level  event_key  match_number  \
key                                                                       
2020wasno_f1m1  2020-03-01 15:18:36          f  2020wasno             1   
2020wasno_f1m2  2020-03-01 15:56:12          f  2020wasno             2   
2020wasno_qf1m1 2020-03-01 13:03:52         qf  2020wasno             1   

                   post_result_time      predicted_time  set_number  \
key                                                                   
2020wasno_f1m1  2020-03-01 15:24:06 2020-03-01 15:20:22           1   
2020wasno_f1m2  2020-03-01 15:59:33 2020-03-01 15:57:20           1   
2020wasno_qf1m1 2020-03-01 13:07:43 2020-03-01 13:04:05           1   

                               time winning_alliance     red1     red2  \
key                                                                      
2020wasno_f1m1  2020-03-01 15:06:00              red  frc4911  frc2910   
2020wasno_f1m2  2020-03-01 15:13:00              red  frc4911  frc2910   
2020wasno_qf1m1 2020-03-01 13:00:00              red  frc4911  frc2910   

                    red3    blue1    blue2    blue3  red_score  blue_score  
key                                                                         
2020wasno_f1m1   frc4173  frc2930  frc2976  frc4918        148         136  
2020wasno_f1m2   frc4173  frc2930  frc2976  frc4918        163         137  
2020wasno_qf1m1  frc4173  frc4512   frc949  frc4131        193         122

There is a significant difference between the space and matches dataframes, besides the fact that they have different columns. Let's look at the first few rows of the space dataframe again.

In [56]:
space_df.head(3)

Company Name                                           Location  \
0       SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
1         CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2       SpaceX                      Pad A, Boca Chica, Texas, USA   

                       Datum                                        Detail  \
0  2020-08-07 05:12:00+00:00  Falcon 9 Block 5 | Starlink V1 L9 & BlackSky   
1  2020-08-06 04:01:00+00:00           Long March 2D | Gaofen-9 04 & Q-SAT   
2  2020-08-04 23:57:00+00:00            Starship Prototype | 150 Meter Hop   

  Status Rocket  Rocket Status Mission  
0  StatusActive   50.00        Success  
1  StatusActive   29.75        Success  
2  StatusActive     NaN        Success

The far left column of the space dataframe contains integers, with the first row having integer 0, the second row having integer 1, and so on. They appear to be row numbers. Unlike all other columns, they are displayed using a bold font.

The matches dataframe also has a column on the far left that is displayed using a bold font, but it does not contain integers. It contains a string with a code that identifies the match, formatted like this: *{year}{event_code}{comp_level_code}m{match_number}*. The column also has a name: *key*. By the way, the syntax of the *key* values has nothing to do with Pandas. This data was retrieved from [The Blue Alliance](https://www.thebluealliance.com/) website, which uses a this key value to uniquely identify every FRC match that occurs at every competition throughout the world.

The column on the far left is a special column. It is called the *index* and it can contain integers, strings, data values or other data types. If we don't tell Pandas how to create the index when we create a dataframe, Pandas will create an integer index, where each index value is the row number. But when the matches dataframe was created, Pandas was told to use *The Blue Alliance's* match key for the index. (If you would like to see how the matches datframe was created, check out the *convert_json_to_df.ipynb* notebook in the *setup* subfolder.)

#### 2. Indexes are Objects
Here is one way to think about indexes: just like every column has a unique name, every row has a unique name. The index is a column that contains the row names.

In [58]:
# Viewing column and row names for the matches dataframe.
print("Matches Dataframe Column Names:\n", matches.columns)
print("\nMatches Dataframe Row Names (first 10):\n", matches.index[:10])

Matches Dataframe Column Names:
 Index(['actual_time', 'comp_level', 'event_key', 'match_number',
       'post_result_time', 'predicted_time', 'set_number', 'time',
       'winning_alliance', 'red1', 'red2', 'red3', 'blue1', 'blue2', 'blue3',
       'red_score', 'blue_score'],
      dtype='object')

Matches Dataframe Row Names (first 10):
 Index(['2020wasno_f1m1', '2020wasno_f1m2', '2020wasno_qf1m1',
       '2020wasno_qf1m2', '2020wasno_qf2m1', '2020wasno_qf2m2',
       '2020wasno_qf3m1', '2020wasno_qf3m2', '2020wasno_qf3m3',
       '2020wasno_qf4m1'],
      dtype='object', name='key')


Look closely at the output from the cell above. First of all, we can easily view all column or row names with the `DataFrame.columns` and `DataFrame.index` attributes. But look closer. See how the printed output starts with `Index([...`? The columns and index are themselves an object with a special datatype. We'll prove it.

In [59]:
# Index object types
print("Index object type:\t\t", type(matches.index))
print("Columns object type:\t\t", type(matches.columns))
print("Object type of a regular column:", type(matches.red_score))

Index object type:		 <class 'pandas.core.indexes.base.Index'>
Columns object type:		 <class 'pandas.core.indexes.base.Index'>
Object type of a regular column: <class 'pandas.core.series.Series'>


Both the index and columns are the same object type: `Index`, which is different than the object type of a regular column like *red_score* (object type is `Series`). This means the index will behave differently than a regular column.

#### 3. Reason for Having Indexes - The `.loc()` Method
Why do Pandas dataframes have an index? The size of a Pandas dataframe is limited only by the amount of memory on your computer.  Pandas dataframes can have millions of rows -- that's not hyperbole. The designers of Pandas wanted to be able to extract data from dataframes quickly, regardless of the size of the dataframe. The `Index` objects make that possible. If we extract data using index values and column names, we can extract data just as quickly from large datasets as we can from small datasets.

There is a caveat. For indexing (i.e., extracting data) to be fast on large dataframes, each row and column must have a unique index value. Pandas will allow duplicate index values, but some operations will be slower, and there could be other problems. The best practice is to ensure each row and column has a unique index value. Your choice of index will depend on the data and how you want to use it.

#### 4. Extracting Data with Index Values
We've already used the method for extracting data with index values. It's the `.loc` method back in section D.4. Here's a review:

In [60]:
space_df.loc[0:3, "Company Name":"Datum"]

Company Name                                           Location  \
0       SpaceX         LC-39A, Kennedy Space Center, Florida, USA   
1         CASC  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2       SpaceX                      Pad A, Boca Chica, Texas, USA   
3    Roscosmos       Site 200/39, Baikonur Cosmodrome, Kazakhstan   

                       Datum  
0  2020-08-07 05:12:00+00:00  
1  2020-08-06 04:01:00+00:00  
2  2020-08-04 23:57:00+00:00  
3  2020-07-30 21:25:00+00:00

When we use `.loc()` on the space dataframe, it appears that `.loc()` uses row numbers. But that's because the index for the space dataframe uses a simple integer index. The same technique will *not* work on the matches dataframe.

In [90]:
# This code generates an error
matches.loc[0:3, "red1":"blue3"]

TypeError: cannot do slice indexing on Index with these indexers [0] of type int

At the bottom of the long error message, you should see a statement indicating that Pandas is unable to use integers to index this dataframe. The `.loc()` method requires us to pass index values.

In [63]:
# Table of teams in quarterfinal matches.
matches.loc["2020wasno_qf1m1":"2020wasno_qf4m3", "red1":"blue3"]

red1     red2     red3    blue1    blue2    blue3
key                                                                  
2020wasno_qf1m1  frc4911  frc2910  frc4173  frc4512   frc949  frc4131
2020wasno_qf1m2  frc4911  frc2910  frc4173  frc4512   frc949  frc4131
2020wasno_qf2m1  frc4089  frc4915  frc4309  frc1318  frc4513  frc7461
2020wasno_qf2m2  frc4089  frc4915  frc4309  frc1318  frc4513  frc7461
2020wasno_qf3m1  frc2522  frc3826  frc8032  frc1778  frc1899   frc492
2020wasno_qf3m2  frc2522  frc3826  frc8032  frc1778  frc1899   frc492
2020wasno_qf3m3  frc2522  frc3826  frc8032  frc1778  frc1899   frc492
2020wasno_qf4m1  frc4681  frc3070  frc2412  frc2930  frc2976  frc4918
2020wasno_qf4m2  frc4681  frc3070  frc2412  frc2930  frc2976  frc4918

##### More Information
* [The loc Function](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#selection-by-label)

#### 5. Extracting Data with Row and Column Numbers -- Using `.iloc()`
There may be times when you don't care about the index or column name. For example, you just want the value from the 5th column of the 17th row. Pandas makes that easy.

In [64]:
# Getting value from specific row and column
# Remember, 1st row and 1st column are at index 0
matches.iloc[16, 4]

Timestamp('2020-02-29 13:20:48')

The `.iloc()` method works similarly to `.loc()`, but it only takes integer row and column numbers. Here's another example:

In [65]:
matches.iloc[30:35, :-9:-1]

blue_score  red_score    blue3    blue2    blue1     red3  \
key                                                                         
2020wasno_qm27          82         79  frc3826  frc8059  frc4309  frc2930   
2020wasno_qm28          56        105  frc4180  frc8032  frc4681   frc949   
2020wasno_qm29          49         89  frc8248  frc4205  frc2976  frc4512   
2020wasno_qm3           12         65  frc4309  frc1294  frc4512  frc4513   
2020wasno_qm30          49         82  frc3268  frc7627  frc5588  frc4918   

                   red2     red1  
key                               
2020wasno_qm27  frc4173  frc7118  
2020wasno_qm28  frc1778  frc1318  
2020wasno_qm29  frc4682  frc2412  
2020wasno_qm3   frc4173  frc4918  
2020wasno_qm30  frc3070  frc4089

The `.iloc()` method accepts list-style slices. In the example above, we selected a range of rows and the last eight columns, but in reverse order.

##### More Information
* [The iloc Function](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#selection-by-position)

#### 6. Index Exercises

##### Ex III.13
For the `grads` dataframe, use the `.set_index()` method to make the *Major_code* column the index. Refer to the [documentation for the `set_index()` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.set_index.html?highlight=set_index#pandas.DataFrame.set_index). Either assign the modified dataframe to a new variable, or use the *inplace* parameter.

In [109]:
# Ex II
grads.set_index("Major_code", inplace = True)

KeyError: "None of ['Major_code'] are in the columns"

##### Ex III.14
With the modified `grads` dataframe from exercise III.13, sort the dataframe in ascending order by the index. Use the `.sort_index()` method ([see the documentation here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_index.html?highlight=sort_index#pandas.DataFrame.sort_index)). Next, extract all columns from *Major_code* to *Employed*, but limit rows to those with major codes between 4000 and 4999.

In [122]:
# Ex III.14
sorted_grads = grads.sort_index()
sorted_grads.loc["4000":"4999", "Rank":"Employed"]

Rank                                          Major     Total  \
Major_code                                                                  
4000         110                MULTI/INTERDISCIPLINARY STUDIES   12296.0   
4001         117        INTERCULTURAL AND INTERNATIONAL STUDIES   24650.0   
4002         104                             NUTRITION SCIENCES   18909.0   
4005          53               MATHEMATICS AND COMPUTER SCIENCE     609.0   
4006          55            COGNITIVE SCIENCE AND BIOPSYCHOLOGY    3831.0   
4007         132              INTERDISCIPLINARY SOCIAL SCIENCES    9916.0   
4101         140  PHYSICAL FITNESS PARKS RECREATION AND LEISURE  125074.0   
4801         136               PHILOSOPHY AND RELIGIOUS STUDIES   54814.0   
4901         159               THEOLOGY AND RELIGIOUS VOCATIONS   30207.0   

                Men    Women                       Major_category  ShareWomen  \
Major_code                                                                      
4000         2817.0   9479.0                    Interdisciplinary    0.770901   
4001         8575.0  16075.0            Humanities & Liberal Arts    0.652130   
4002         2563.0  16346.0                               Health    0.864456   
4005          500.0    109.0              Computers & Mathematics    0.178982   
4006         1667.0   2164.0               Biology & Life Science    0.564866   
4007         2337.0   7579.0                       Social Science    0.764320   
4101        62181.0  62893.0  Industrial Arts & Consumer Services    0.502846   
4801        31967.0  22847.0            Humanities & Liberal Arts    0.416810   
4901        18616.0  11591.0            Humanities & Liberal Arts    0.383719   

            Sample_size  Employed  
Major_code                         
4000                128      9821  
4001                184     18824  
4002                118     13217  
4005                  7       559  
4006                 25      2741  
4007                 95      7444  
4101               1014    103078  
4801                375     40157  
4901                310     24202

## H. Making your own DataFrames from Scratch
The two most common ways to construct a `DataFrame` object from core Python objects are the column method and the row method.

For the column method, you pass a dictionary to the `Pandas.DataFrame` constructor.
* Each key of the dictionary is a column name.
* Each value is a list of values that will go in the corresonding column.
* Each list must be the same length.

In [123]:
# Column method for creating a dataframe
x = pd.DataFrame({'x': [1, 2, 3], 'y': [3, 4, 5]})
x

x  y
0  1  3
1  2  4
2  3  5

For the row method, you pass a list of dictionaries to the `Pandas.DataFrame` constructor.
* Each dictionary corresponds to one row of the dataframe.
* Each dictionary key is a column name and each dictionary value will be the value for the corresponding row and column.
* This method is more fault tolerant. If a dictionary is missing a column value, Pandas will not throw an error, but will insert a `NaN` value.

In [124]:
x2 = pd.DataFrame([{"x": 10, "y": 30},
                   {"x": 20, "y": 40, "z": 100},
                   {"x": 30, "y": 50}])
x2

x   y      z
0  10  30    NaN
1  20  40  100.0
2  30  50    NaN

##### More Information
* [Series and Dataframe Creation](https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html#object-creation)

## I. Modifying Data in Dataframes
So far we've only been reading data from Pandas DataFrames. But we can change dataframes. They are mutable.

In [125]:
# You can also assign a dict to a row of a DataFrame
x.iloc[1] = {'x': 9, 'y': 99}
x

x   y
0  1   3
1  9  99
2  3   5

In [126]:
x = x.append({'x': 5, 'y': 9}, ignore_index = True)
x

x   y
0  1   3
1  9  99
2  3   5
3  5   9

In [127]:
x['z'] = [1, 2, 3, 4]
x

x   y  z
0  1   3  1
1  9  99  2
2  3   5  3
3  5   9  4

In [128]:
z = pd.DataFrame({'x': [1, 2, 3], 'y': [3, 4, 5], 'z': [45, 45, 56]})
z

x  y   z
0  1  3  45
1  2  4  45
2  3  5  56

In [129]:
#You can append 2 dataframes together
x = x.append(z)
x

x   y   z
0  1   3   1
1  9  99   2
2  3   5   3
3  5   9   4
0  1   3  45
1  2   4  45
2  3   5  56

#### 1. Dataframe Creation Exercises

##### Ex III.15
Use list comprehensions to create three lists each with 20 numbers. The first list should contain integers ranging from 0 to 19. The second should contain the squares of the numbers in the first list, and the third list should contain the cubes. Create a dataframe with three columns, where each column contains one of the lists. Use descriptive, logical names for the columns.

In [148]:
# Ex III.15
num = pd.DataFrame({'normal':[num for num in range(0, 20)], 'square': [num ** 2 for num in range(0, 20)], 'cube': [num ** 3 for num in range(0, 20)]})
num

normal  square  cube
0        0       0     0
1        1       1     1
2        2       4     8
3        3       9    27
4        4      16    64
5        5      25   125
6        6      36   216
7        7      49   343
8        8      64   512
9        9      81   729
10      10     100  1000
11      11     121  1331
12      12     144  1728
13      13     169  2197
14      14     196  2744
15      15     225  3375
16      16     256  4096
17      17     289  4913
18      18     324  5832
19      19     361  6859

## J. Grouping and Aggregating Data

Until now, we've been using only a fraction of Pandas' capabilities. Pandas Pandas can also be used to transform and analyze data.

Suppose we want to know the average alliance score for the entire competition. Since scores are contained in either the *blue_score* or *red_score* column. We could use the `.mean()` method to calculate the average of each column and then average those two values.

In [149]:
# Calculating the mean of the red and blue score columns
print("Blue alliance mean score:\t",
      round(matches.blue_score.mean(), 1))
print("Red alliance mean score:\t", round(matches.red_score.mean(), 1))
print("Overall mean score:\t\t",
      round((matches.red_score.mean() + matches.blue_score.mean()) / 2, 2))

Blue alliance mean score:	 87.9
Red alliance mean score:	 90.1
Overall mean score:		 89.0


With Pandas, there is usually more than one way to do something, and this example is no exception. For example, we could also add a column to the dataframe that contains the mean of of the red and blue alliance scores for each match, and just take the mean of that column.

In [150]:
# Creating a new mean_score column
matches["mean_score"] = (matches["red_score"] + matches["blue_score"]) / 2    # line 1
print("Overall mean score:\t", round(matches.mean_score.mean(), 1)) # line 2

Overall mean score:	 89.0


Pay close attention to line 1 -- there is a lot going on in that line.
* First, we're able to create a new column named *mean_score* simply by referencing the column and assigning something to it, e.g., `matches["mean_score"] = ...`.
* Line 1 is also using element-wise calculations. This is an important concept, so we'll cover it in detail.
    * The expressions `matches["red_score"]` and `matches["blue_score"]` both return a Pandas `Series` object.
    * Mathematical operators like `+` behave differently with `Series` objects than they behave with other data types, such as Python lists. See below for an example.

In [151]:
# Using '+' with Python lists and Pandas Series
list1 = [1, 2, 3]
list2 = [10, 20, 30]
print("Using '+' with Python lists:\t", list1 + list2)
series1 = pd.Series([1, 2, 3])
series2 = pd.Series([10, 20, 30])
series3 = series1 + series2
print("\nUsing '+' with Pandas series:")
series3

Using '+' with Python lists:	 [1, 2, 3, 10, 20, 30]

Using '+' with Pandas series:


0    11
1    22
2    33
dtype: int64

Adding two Python lists concatenates the two lists into a longer list. But adding two `Series` objects causes the first elements of each series to be summed, as well as the second elements, etc. This is what we mean by element-wise operations.

So when we added `matches["red_score"]` and `matches["blue_score"]` and divided the result by two, for every row in the `matches` dataframe, we took the red score and the blue score, added them together, divided the sum by two, and put the result in a new column called *mean_score*. The following cell shows the results.

In [152]:
matches[["blue_score", "red_score", "mean_score"]].head()

blue_score  red_score  mean_score
key                                               
2020wasno_f1m1          136        148       142.0
2020wasno_f1m2          137        163       150.0
2020wasno_qf1m1         122        193       157.5
2020wasno_qf1m2          95        140       117.5
2020wasno_qf2m1         134         89       111.5

Now suppose we were interested in the average score for each level of competition. That is, suppose we wanted an average score for qualification matches, another average score for quarter-finals, another for semi-finals, and so on. We could use the indexing techniques to extract a dataframe containing only the rows corresponding to each level, and then calculate a mean from each separate dataframe, but that would be a lot of work. It's easier to do something like this:

In [153]:
matches.groupby("comp_level").agg({"mean_score": "mean"})

mean_score
comp_level            
f           146.000000
qf          117.277778
qm           81.621622
sf          133.375000

This result is actually quite remarkable. One line of code split our data into groups by competition level and then calculated a mean score for each level.

First, the `groupby()` method created a Pandas `GroupBy` object, where rows are split into groups based on the content of the *comp_level* column. We can extract the individual groups with the `.get_group()` method. For example, the following line extracts the finals matches.

In [154]:
matches.groupby("comp_level").get_group("f")

actual_time comp_level  event_key  match_number  \
key                                                                      
2020wasno_f1m1 2020-03-01 15:18:36          f  2020wasno             1   
2020wasno_f1m2 2020-03-01 15:56:12          f  2020wasno             2   

                  post_result_time      predicted_time  set_number  \
key                                                                  
2020wasno_f1m1 2020-03-01 15:24:06 2020-03-01 15:20:22           1   
2020wasno_f1m2 2020-03-01 15:59:33 2020-03-01 15:57:20           1   

                              time winning_alliance     red1     red2  \
key                                                                     
2020wasno_f1m1 2020-03-01 15:06:00              red  frc4911  frc2910   
2020wasno_f1m2 2020-03-01 15:13:00              red  frc4911  frc2910   

                   red3    blue1    blue2    blue3  red_score  blue_score  \
key                                                                         
2020wasno_f1m1  frc4173  frc2930  frc2976  frc4918        148         136   
2020wasno_f1m2  frc4173  frc2930  frc2976  frc4918        163         137   

                mean_score  
key                         
2020wasno_f1m1       142.0  
2020wasno_f1m2       150.0

Next, the `.agg()` method calculates the mean of the values in each group's *mean_score* column. The string "mean" refers to an aggregate function, which is a function that calculates a single number from many different numbers. Pandas provides numerous aggregate functions, including *sum*, *size*, *count*, *std* (standard deviation), *var* (variance), *min*, and *max*. We can add additional columns that calculate different summary statistics for each group. The following example adds a column with the earliest match start time for each group.

In [155]:
matches.groupby("comp_level").agg({"actual_time": "min", "mean_score": "mean"})

actual_time  mean_score
comp_level                                
f          2020-03-01 15:18:36  146.000000
qf         2020-03-01 13:03:52  117.277778
qm         2020-02-29 11:04:24   81.621622
sf         2020-03-01 14:30:03  133.375000

##### More Information
* [Groub by: Split-Apply-Combine](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html)

#### 1. Aggregation Exercises

##### Ex III.16
Using the `grads` dataframe, calculate the average unemployment rate and the standard deviation of the median salaries for each category of majors.

In [161]:
# Ex III.16
grads.groupby("Major_category").agg({"Median": "std"})

Median
Major_category                                   
Agriculture & Natural Resources       6935.416354
Arts                                  7223.164621
Biology & Life Science                4528.912006
Business                              7774.052832
Communications & Journalism           1000.000000
Computers & Mathematics               5108.691346
Education                             3892.728263
Engineering                          13626.079747
Health                                5776.460854
Humanities & Liberal Arts             3393.032076
Industrial Arts & Consumer Services   7290.829204
Interdisciplinary                             NaN
Law & Public Policy                   9066.421565
Physical Sciences                     8251.659766
Psychology & Social Work              5381.914158
Social Science                        4750.555523

## IV. Quiz
Answer the following questions by typing the answers as comments in the code block below each question.

**#1.** Of the data file formats we have studied so far, which one can be viewed in a standard text editor *and* can contain nested data?

In [ ]:
# 1. 
# .CSV files

**#2.** Tab separated files (TSV) are very similar to comma separated files. Tab characters, '\t' are used to separate columns instead of commas. This format can be useful if the data itself contains commas.

Review the [documentation for the Python Standard Library's `csv` module](https://docs.python.org/3/library/csv.html?highlight=csv#module-csv). Can this module be used for reading tab TSV files? If so, write out the code that could be used to load a TSV file.

In [162]:
# 2.
# import csv
# with open('name_of_file.tsv') as tsvfile:
#    read_tsvfile = csv.reader(tsvfile)

**#3.**  This code will throw an error. Why? How do we fix it?

```python
space_df["Company Name", "Datum", "Detail"]
```

In [163]:
# 3. 
# There needs to be another square braket around the columns!

**#4.** What are two differences between a Python list and a Pandas Series?

In [164]:
# 4.
# All elements of a Series must have the same data type and the contents of a Panda Series are store in memory more
# efficiently than lists resulting in faster fcalculations

**#5.** What is the difference between the `.loc()` and `.iloc()` methods?

In [165]:
# 5.
# .iloc takes integer row and column numbers while .loc takes the row numbers and column names.

**#6.** When are tick marks required when using the `.query()` methods? How are tick marks different than quotation marks?

In [166]:
# 6.
# Tick marks are required when using special characters. Tick marks are in the top right of your keyboard while quotation
# marks are on the bottom right.

**#7.** Suppose we have a dataframe `small_df` with two numeric columns, *col1* and *col2*.

|  |col1|col2|
|--|----|----|
|0 |1   |2   |
|1 |10  |20  |
|2 |100 |NaN |

Suppose we run the following line of code?
```python
small_df["col3"] = small_df["col1"] * small_df["col2"]
```
What values will *col3* contain? Review the [user guide](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html#calculations-with-missing-data) to see how *NaN* values are handled within mathematical calculations.

In [167]:
# 7.
# 2 200 1

**#8.** Many Pandas methods, including several used earlier in this notebook, have an *inplace* parameter. What does this parameter do? Review the API documentation for some of the functions that we have used if you are not sure.

In [168]:
# 8.
#  The parameter modifies the dataframe directly and does not create a new object.

**#9.** We discussed two different techniques for creating a Pandas dataframe from Python data structures. One technique used a dictionary of lists, and the other used a list of dictionaries. Which one is the column-centric approach, and which is the row-centric approach?

In [169]:
# 9.
# Dictonary of lists is column-centric while list of dictionaries is row-centric.

**#10.** Review the *`setup/convert_json_to_df.ipynb`* notebook. This notebook was used to create the matches dataframe that we used in this session. The dataframe was created using data in the *`setup/matches.json`* file. Review the notebook and JSON file and try to figure out how it works. Refer to Python or Pandas documentation if any of the syntax looks unfamiliar.

There is a line of code in the notebook that merges two dataframes. Copy the line of code into your answer to this question.

In [172]:
# 10.
# matches = match_meta.merge(match_teams, on=\"key\")

**#11.** The answer to question #10 contains a method that takes more than one argument. Explain what each argument does. You can easily find the documentation for the method by typing it into the search field on the Pandas documentation website (upper left).

In [173]:
# 11.
# the on=key argument basically means that the dataframes are merging on the indexes.

## V. Save Your Work
Once you have completed the exercises, save a copy of the notebook outside of the git repository (outside of the *pyclass_frc* folder). Include your name in the file name. Follow instructions from your instructor to get feedback on your work.

## VI. Concept and Terminology Review
You should be able to define the following terms or describe the concept.
* CSV files
* TSV files
* The csv module from the Python Standard Library
* Creating dataframes, including both the row and column approach
* Pandas `DataFrame` and `Series` datatypes
* Pandas `.head()` and `.tail()` methods
* `Dataframe.shape` attribute
* Selecting dataframe columns
* Selecting rows from a Series
* Pandas `.loc` and `.iloc` methods
* Boolean indexing
* Pandas `.query()` method
* Element-wise operations
* The `.isin()` and `isna()`, and `.notna()` methods
* *NaN* values
* Column datatypes
* Pandas `.to_datetime()` and `.astype()` methods
* Pandas statistical functions
* Pandas Indexes
* The `DataFrame.set_index()` method
* Modifying data in dataframes
* Adding new columns to a dataframe
* Grouping and Aggregating data
* Sorting the dataframe by a column or by the index (`.sort_index()` and `sort_values()` methods)
* Merging Pandas dataframes

### Notes for later

[Table of Contents](../../index.ipynb)